pytorch_lightning/core/lightning.py

```diff
@@ -1591,8 +1595,13 @@ def to_torchscript(
                 # if no example inputs are provided, try to see if model has example_input_array set
                 if example_inputs is None:
                     example_inputs = self.example_input_array
+                # dicts are not supported, so show the user an error; not raising an error to show the original error
```

Comment:
I don't understand this comment, it is saying there is an error but there is not an error, what is it?
can we just remove the comment? the code should speak for itself

Reply:
the inputs must be a tensor or tuple of tensors. IMO a better way to handle this is by wrapping the input tensor into a tuple the checking whether each element in the tuple is an instance of Tensor or not.

Reply:
Also with a test for the same.

Reply:
@awaelchli good catch, that doesn't make sense indeed.
@rohitgr7 I think `trace()` already does internally this wrapping of a `torch.Tensor` in a tuple, so I don't think we have to add that again on Lightning's side?

Reply:
I changed the comment to hopefully make more sense

Reply:
The reason I asked for this is to check for input whether they are tensors or not. For now it only checks whether it's an instance of `Mapping` or not but won't throw any error from PL when it's a custom datatype derived from some custom class. Either we should remove this check too and let torchscript handle this or check these valid inputs here itself.

Reply:
not sure about all these custom error messages. I mean, doesn't pytorch provide a good message if you give it unsupported arguments?

Reply:
yep it does. Then let's remove any such checks and let torchscript handle this. @NumesSanguis

Reply:
@awaelchli @rohitgr7 I just added the following line to The BoringModel to see what the error message is:
```python
model.to_torchscript(method='trace', example_inputs={"sample": torch.Tensor([0.2] * 32)})
```

Error: `AttributeError: 'dict' object has no attribute 'to'`

<details><summary>Full Traceback (CLICK ME)</summary>
<p>

```python
/usr/local/lib/python3.6/dist-packages/pytorch_lightning/core/lightning.py in to_torchscript(self, file_path, method, example_inputs, **kwargs)
   1593                     example_inputs = self.example_input_array
   1594                 # automatically send example inputs to the right device and use trace
-> 1595                 torchscript_module = torch.jit.trace(func=self.eval(), example_inputs=example_inputs.to(self.device),
   1596                                                      **kwargs)
   1597             else:

AttributeError: 'dict' object has no attribute 'to'
```

</p>
</details>

Do you think this error message is clear enough by itself? I think if the user provides `example_inputs` as argument, it is already clear enough. However, if it is automatically taken from `self.example_input_array`, which can be set anywhere in the script, only this message might be confusing?

If you both think that for both cases this error message is already clear enough, I'll remove it.

Reply:
> /usr/local/lib/python3.6/dist-packages/pytorch_lightning/core/lightning.py in to_torchscript(self, file_path, method, example_inputs, **kwargs)
   1593                     example_inputs = self.example_input_array
   1594                 # automatically send example inputs to the right device and use trace
-> 1595                 torchscript_module = torch.jit.trace(func=self.eval(), example_inputs=example_inputs.to(self.device),
   1596                                                      **kwargs)
   1597             else:
AttributeError: 'dict' object has no attribute 'to'

@NumesSanguis you seem to be using the master. Try with the changes made in this PR after removing the check for Mapping.

Reply:
@rohitgr7 We wanted to know the error message given by `torch.jit.trace()` right? That didn't change between master and this request. So my question is if this error by `torch.jit.trace()` is clear enough (`AttributeError: 'dict' object has no attribute 'to'`). If yes, there is no need to add this one: `log.error("`example_inputs` should be a Tensor or a tuple of Tensors, but got {type(example_inputs)}.")`.

If the above error message is not enough, we should keep the `log.error()` from PL's side.

Reply:
@rohitgr7 Here is the Traceback from this PR:

<details><summary>Full Traceback (CLICK ME)</summary>
<p>

```python
--------------------------------------------------------------------------------
DATALOADER:0 TEST RESULTS
{'fake_test_acc': tensor(9.7542e-15)}
--------------------------------------------------------------------------------

`example_inputs` should be a Tensor or a tuple of Tensors, but got <class 'dict'>.



---------------------------------------------------------------------------

AttributeError                            Traceback (most recent call last)

<ipython-input-11-1f9f6fbe4f6c> in <module>()
----> 1 test_x(tmpdir)

10 frames

/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py in linear(input, weight, bias)
   1670         if any([type(t) is not Tensor for t in tens_ops]) and has_torch_function(tens_ops):
   1671             return handle_torch_function(linear, tens_ops, input, weight, bias=bias)
-> 1672     if input.dim() == 2 and bias is not None:
   1673         # fused op is marginally faster
   1674         ret = torch.addmm(bias, input, weight.t())

AttributeError: 'dict' object has no attribute 'dim'
```

</p>
</details>

Reply:
@rohitgr7 Aaah I see what you mean now. The previous error was indeed not about `torch.jit.trace()`, but that a dict has no `.to()` function :')

Reply:
The [above error](https://github.com/PyTorchLightning/pytorch-lightning/pull/4360#discussion_r513210011) is from the model and not the torchscript because it doesn't handle dict inputs.

here `torch.jit.trace` doesn't throw any error with dict as long as all the values are of same type and it is accepted by the `forward`. Also [check this](https://github.com/pytorch/pytorch/issues/16847#issuecomment-485863129). Not super familiar with torchscript but will check what's happening here and why the docs says that it accepts only tensor or tuple of tensors.

Reply:
@rohitgr7 Now you mention it, it is indeed not from TorchScript. So actually it does work at TorchScript's side (but they say they don't support it), but Lightning (who says it's supported) is at fault.

In relation to this PR, should we just leave the dict logger error here until the issue has been identified and fixed (it doesn't stop the execution after all)? Then this PR can be merged for now, which is already better than the current `.to_torchscript(method='trace')` implementation.

Reply:
give me sometime. Need to check what's the actual issue here. Is there something wrong from pytorch side or we are doing something wrong here? In the meantime can you open an issue on pytorch forums if possible? Maybe we can get a quick response there :) Would be good to resolve all issues in this PR itself to avoid any issues in the future related to `to_torchscript`.

Reply:
This issue is already in the master, since this pull request was already merged: https://github.com/PyTorchLightning/pytorch-lightning/issues/4140
This pull request is just to add some quality of life changes to the previous one. If we merge this one, it's much easier for other people to reproduce this issue, because they will have the same error output (this pull request does not add a new problem, just 1 step closer to solving it).

We can just keep the original issue (https://github.com/PyTorchLightning/pytorch-lightning/issues/4140) open, and discuss this issue there, as it would be more easy to find compared to this comment thread. Then we can point a PyTorch forum issue to there. A new pull request can then target that specific dict improvement (which might be very deep), instead of making this PR huge.

Honestly, I would like to make all parts work nicely, but I'm not affected by the dict issue, and I already spend too much time on this pull request. The previous pull request already added everything needed for my use case, but this pull request is just an extra to make the previous one a little bit less rough.

Reply:
Ok cool. Then let's remove the check for Mapping and merge this one since it doesn't throw any error with dict :)

Reply:
@rohitgr7 Thanks. The logger error has been removed :)

Reply:
@rohitgr7 I put a summary of the Dict issue here: https://github.com/PyTorchLightning/pytorch-lightning/issues/4140#issuecomment-718381421
which should make the discussion a bit more visible for others.